listenbrainz/listenstore/timescale_listenstore.py

```diff
@@ -208,94 +246,158 @@ def insert(self, listens):
 
         conn.commit()
 
-        # So update the listen counts of the users cached in brainzutils cache.
-        for _, _, user_name in inserted_rows:
-            user_key = "{}{}".format(self.ns + REDIS_TIMESCALE_USER_LISTEN_COUNT, user_name)
-            cached_count = cache.get(user_key, decode=False)
-            if cached_count:
-                cache.increment(user_key)
+        # update the listen counts and timestamps for the users
+        user_timestamps = {}
+        user_counts = defaultdict(int)
+        for ts, _, user_name in inserted_rows:
+            if listen.user_name in user_timestamps:
+                if ts < user_timestamps[listen.user_name][0]:
+                    user_timestamps[listen.user_name][0] = ts
+                if ts > user_timestamps[listen.user_name][1]:
+                    user_timestamps[listen.user_name][1] = ts
+            else:
+                user_timestamps[listen.user_name] = [ts, ts]
 
-        # Invalidate cached data for user
-        for user_name in user_names:
-            cache.delete(self.ns + REDIS_USER_TIMESTAMPS % user_name)
+            user_counts[user_name] += 1
+
+        for user_name in user_counts:
+            cache._r.incrby(cache._prep_key(
+                REDIS_USER_LISTEN_COUNT + user_name), user_counts[user_name])
+
+        for user in user_timestamps:
+            self.update_timestamps_for_user(
+                user, user_timestamps[user][0], user_timestamps[user][1])
 
         return inserted_rows
 
-    def fetch_listens_from_storage(self, user_name, from_ts, to_ts, limit, order, time_range):
+    def fetch_listens_from_storage(self, user_name, from_ts, to_ts, limit, order):
         """ The timestamps are stored as UTC in the postgres datebase while on retrieving
             the value they are converted to the local server's timezone. So to compare
             datetime object we need to create a object in the same timezone as the server.
 
+            If neither from_ts nor to_ts is provided, the latest listens for the user are returned.
```

Comment:
how many is "the latest" ?

Reply:
That depends on what you pass in for limit. I don't see your question here.